app/src/main/java/com/android/partagix/model/Database.kt

```diff
@@ -388,6 +391,117 @@ class Database(database: FirebaseFirestore = Firebase.firestore) {
     onSuccess(user)
   }
 
+  /**
+   * Retrieve all comments that a user has received, both as a owner than as an loaner
+   *
+   * @param userId the user's id
+   * @param onSuccess the function to return a List containing pairs (comment's author name,
+   *   comment)
+   */
+  fun getComments(userId: String, onSuccess: (List<Pair<String, String>>) -> Unit) {
+    val ret = mutableListOf<Pair<String, String>>()
+
+    getLoans { loans ->
+      for (loan in loans) {
+
+        if (loan.state == LoanState.FINISHED // only finished loans
+        &&
+            loan.idOwner == userId // if the user is the owner,
+            &&
+            loan.reviewOwner.toDouble() != 0.0 // that has been reviewed,
+            &&
+            loan.commentOwner != "") { // and received a comment
+
+          getUser(loan.idLoaner, onNoUser = {}) { user ->
+            ret.add(Pair(user.name, loan.commentOwner))
+          }
+        } else if (loan.state == LoanState.FINISHED // only finished loans
+        &&
+            loan.idLoaner == userId // if the user is the loaner,
+            &&
+            loan.reviewLoaner.toDouble() != 0.0 // that has been reviewed
+            &&
+            loan.commentLoaner != "") { // and received a comment
+
+          getUser(loan.idOwner, onNoUser = {}) { user ->
+            ret.add(Pair(user.name, loan.commentLoaner))
+          }
+        }
+      }
+    }
+
+    onSuccess(ret)
+  }
+
+  /**
+   * Retrieve all ranks that a user has received, both as a owner than as an loaner, compute the
+   * average rank, and store it in the user's rank
+   *
+   * @param idUser the user's id
+   */
+  fun newAverageRank(idUser: String) {
+
+    var rankSum = 0.0
+    var rankCount = 0L
+
+    getLoans { loans ->
+      for (loan in loans) {
+
+        if (loan.state == LoanState.FINISHED // only finished loans
+        &&
+            loan.idOwner == idUser // if the user is the owner,
+            &&
+            loan.reviewOwner.toDouble() != 0.0) { // that has been reviewed
+
+          rankSum += loan.reviewOwner.toDouble()
+          rankCount++
+        } else if (loan.state == LoanState.FINISHED // only finished loans
+        &&
+            loan.idLoaner == idUser // if the user is the loaner,
+            &&
+            loan.reviewLoaner.toDouble() != 0.0) { // that has been reviewed
+
+          rankSum += loan.reviewLoaner.toDouble()
+          rankCount++
+        }
+      }
+      val averageRank = rankSum / rankCount.toDouble()
```

Comment:
Calling this function on a user that has 0 loans will result in a division by zero here if I'm not misunderstanding this.
BTW we should return something maybe different than just 0 as average for users that don't have evaluations yet.

Reply:
Well pointed out, I added a condition to avoid dividing by zero.
However as we discussed earlier this week, it's been agreed than 0 would be the default value for people that hasn't been reviewed yet.

Reply:
okay great, let's keep this then